pkg/webhook/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
+// - 'newPlan' some new plan that should be triggered
+func validateUpdate(old, new *kudov1beta1.Instance, ov *kudov1beta1.OperatorVersion) (*string, error) {
+	// PREREQUISITES:
+	newPlan := new.Spec.PlanExecution.PlanName
+	oldPlan := old.Spec.PlanExecution.PlanName
+	newOvRef := new.Spec.OperatorVersion
+	oldOvRef := old.Spec.OperatorVersion
+
+	paramDiff := kudov1beta1.ParameterDiff(old.Spec.Parameters, new.Spec.Parameters)
+	paramDefs := kudov1beta1.GetParamDefinitions(paramDiff, ov)
+	triggeredPlan, err := triggeredPlan(paramDefs, ov)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
+	}
+
+	hasPlan := oldPlan != ""
+	isParameterUpdate := triggeredPlan != nil
+	isUpgrade := newOvRef != oldOvRef
+	isNovelPlan := !hasPlan && newPlan != ""
+	isPlanOverride := hasPlan && newPlan != "" && newPlan != oldPlan
+	isPlanCancellation := hasPlan && newPlan == ""
+
+	// Checking compatibility rules described in the table above:
+	switch {
+	case hasPlan && isParameterUpdate && *triggeredPlan != oldPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: plan '%s' is scheduled (or running) and an update would trigger a different plan '%s'", old.Namespace, old.Name, oldPlan, newPlan)
+	case isUpgrade && hasPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s while a plan '%s' is scheduled (or running) is not allowed", old.Namespace, old.Name, newOvRef, oldPlan)
+	case isUpgrade && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s and triggering new plan '%s' is not allowed", old.Namespace, old.Name, newOvRef, newPlan)
+	case isPlanOverride:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: overriding currently scheduled (or running) plan '%s' with '%s' is not supported", old.Namespace, old.Name, oldPlan, newPlan)
+	case isPlanCancellation:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: cancelling currently scheduled (or running) plan '%s' is not supported", old.Namespace, old.Name, oldPlan)
+	case isParameterUpdate && isUpgrade:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s together with a parameter update triggering '%s' is not allowed", old.Namespace, old.Name, newOvRef, *triggeredPlan)
+	case isParameterUpdate && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: triggering one plan '%s' directly and through parameter update '%s' is not allowed", old.Namespace, old.Name, oldPlan, newPlan)
+	case isParameterUpdate && isPlanOverride:
```

Comment:
I think I still must be missing something here as this case can never happen and there is no todo or help for person who comes in implementing overrides I feel like this is dead code that will be forgotten about. As much as I don't like TODOs if you insist on keeping this in, please add some explanation :)